provd/internal/services/user/user.go

```diff
@@ -28,78 +26,85 @@ const (
+	s := &Service{
+		conn: conn,
 	}
-}
 
-const validSaltChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789./"
+	// Default objects initialization
+	s.accounts = conn.Object(consts.DbusAccountsPrefix, "/org/freedesktop/Accounts")
 
-// generateSalt generates a salt string of the specified length using validSalts characters.
-func generateSalt(length int) (string, error) {
-	salt := make([]byte, length)
-	for i := range salt {
-		index, err := rand.Int(rand.Reader, big.NewInt(int64(len(validSaltChars))))
-		if err != nil {
-			return "", err
-		}
-		salt[i] = validSaltChars[index.Int64()]
+	err := s.accounts.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping default DBus Accounts object")
 	}
-	return string(salt), nil
-}
 
-// hashPassword hashes the given password, returning in the SHA-512 crypt format.
-// A salt can be provided for testing purposes.
-func hashPassword(password string, testSalt *string) (string, error) {
-	if password == "" {
-		return "", status.Errorf(codes.InvalidArgument, "received an empty password")
+	s.hostname = conn.Object(consts.DbusHostnamePrefix, "/org/freedesktop/hostname1")
+
+	err = s.hostname.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping default DBus Hostname object")
 	}
-	var salt string
-	var err error
 
-	if testSalt != nil {
-		salt = *testSalt
-	} else {
-		salt, err = generateSalt(16)
-		if err != nil {
-			return "", err
+	// Applying options, checking for errors in obtaining DBus objects
+	for _, opt := range opts {
+		if err := opt(s); err != nil {
+			return nil, err
 		}
 	}
-	hasher := sha512.New()
-	hasher.Write([]byte(salt + password))
-	hashedBytes := hasher.Sum(nil)
-	hashedPassword := base64.RawStdEncoding.EncodeToString(hashedBytes)
 
-	return fmt.Sprintf("$6$%s$%s", salt, hashedPassword), nil
+	return s, nil
 }
 
 // CreateUser creates a new user on the system.
-func (s *Service) CreateUser(ctx context.Context, req *pb.CreateUserRequest) (*emptypb.Empty, error) {
+func (s *Service) CreateUser(ctx context.Context, req *pb.CreateUserRequest) (_ *emptypb.Empty, err error) {
+	// Get current statichostname
+	var currentHostname string
+	property, err := s.hostname.GetProperty(consts.DbusHostnamePrefix + ".StaticHostname")
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get current hostname: %v", err)
+	}
+
+	currentHostname, ok := property.Value().(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "failed to convert current hostname to string")
+	}
+
+	var userID uint64
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		// Rollback hostname
+		err = s.hostname.Call(consts.DbusHostnamePrefix+".SetStaticHostname", 0, currentHostname, false).Err
+		if err != nil {
+			slog.Error(fmt.Sprintf("failed to rollback hostname: %v", err))
```

Comment:
I tried to apply this when running a service:
- everything that stops the service is then an error and should be logged as an error
- everything that don’t directly stops the services, but is important to the user is then just a warning (otherwise, we would have stopped).

Here, in the recovery, not resetting the hostname is a best effort as a rollback item, but failing to do so is not exactly what makes us stop (it’s the previous error that did), so I would go for just a warning in that case.

Also, you can avoid an undeed warning confusing the error by only trying to rollback if the currentHostame != from the one you request to change it too? (we will still have cases then when we reset even if we didn’t change it, but it’s a little bit more constraints still)